File: Labs/Lab5/Lab5.java
package Lab5;

public class Lab5 {
    //given treenode class
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    //intial method call to start at the root
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    //private helper method for the public isValidBST()
    //checks BST property of the children of node variable
    private boolean isValidBST(TreeNode node, long lower, long upper) {
        //check for empty node
        if (node == null) {
            return true;
        }
        //checks for BST property
        if (node.val <= lower || node.val >= upper) {
            return false;
        }
        if (!isValidBST(node.right, node.val, upper)) {
            return false;
        }
        if (!isValidBST(node.left, lower, node.val)) {
            return false;
        }
        //BST property is present
        return true;
    }

    public static void main(String[] args) {
        Lab5 solution = new Lab5();

        //test case tree 1
        TreeNode node1 = new TreeNode(1);
        TreeNode node3 = new TreeNode(3);
        TreeNode node5 = new TreeNode(5);
        TreeNode node9 = new TreeNode(9);
        TreeNode node8 = new TreeNode(8, node5, node9);
        TreeNode node4 = new TreeNode(4, node3, node8);
        node3.left = node1;

        // Output
        System.out.println(solution.isValidBST(node4));  

        //test case tree 2
        TreeNode node2 = new TreeNode(2);
        TreeNode node3_2 = new TreeNode(3);
        TreeNode node4_2 = new TreeNode(4);
        TreeNode node5_2 = new TreeNode(5);
        TreeNode node6 = new TreeNode(6);
        TreeNode node8_2 = new TreeNode(8, node5_2, node6);
        TreeNode node1_2 = new TreeNode(1, node2, node8_2);
        node2.left = node3_2;
        node2.right = node4_2;

        // Output
        System.out.println(solution.isValidBST(node1_2));
    }
}

File: Labs/Lab5/Lab5.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    # intial method call to start at the root
    def isValidBST(self, root):
        return self._isValidBST(root, float('-inf'), float('inf'))
    
    #helper method part for isValidBST(self, root)
    #checks BST property of the children of node variable
    def _isValidBST(self, node, lower, upper):
        # check for empty node
        if not node:
            return True
        # checks for BST property
        if node.val <= lower or node.val >= upper:
            return False

        if not self._isValidBST(node.right, node.val, upper):
            return False

        if not self._isValidBST(node.left, lower, node.val):
            return False
        # BST property is present
        return True

# Test cases
def test_isValidBST():
    solution = Solution()
    
    # test cast tree 1
    node1 = TreeNode(1)
    node3 = TreeNode(3)
    node5 = TreeNode(5)
    node9 = TreeNode(9)
    node8 = TreeNode(8, node5, node9)
    node4 = TreeNode(4, node3, node8)
    node3.left = node1
    assert solution.isValidBST(node4) == True
    
    # test cast tree 2
    node3 = TreeNode(3)
    node4 = TreeNode(4)
    node5 = TreeNode(5)
    node6 = TreeNode(6)
    node8 = TreeNode(8, node5, node6)
    node2 = TreeNode(2, node3, node4)
    node1 = TreeNode(1, node2, node8)
    assert solution.isValidBST(node1) == False

    print("All test cases pass")

test_isValidBST()
